trin-core/src/utp/stream.rs

```diff
@@ -488,179 +496,249 @@ impl UtpSocket {
+                // If three or more packets are acknowledged past the implicit missing one,
+                // assume it was lost.
+                if extension.iter().count_ones() >= 3 {
+                    self.resend_lost_packet(packet.ack_nr() + 1);
+                    packet_loss_detected = true;
                 }
-            }
 
-            // Three duplicate ACKs mean a fast resend request. Resend the first unacknowledged packet
-            // if the incoming packet doesn't have a SACK extension. If it does, the lost packets were
-            // already resent.
-            if !self.send_window.is_empty()
-                && self.duplicate_ack_count == 3
-                && !packet
-                    .extensions()
-                    .any(|ext| ext.get_type() == ExtensionType::SelectiveAck)
-            {
-                self.resend_lost_packet(packet.ack_nr() + 1);
+                if let Some(last_seq_nr) = self.send_window.last().map(Packet::seq_nr) {
+                    let lost_packets = extension
+                        .iter()
+                        .enumerate()
+                        .filter(|&(_, received)| !received)
+                        .map(|(idx, _)| packet.ack_nr() + 2 + idx as u16)
+                        .take_while(|&seq_nr| seq_nr < last_seq_nr);
+
+                    for seq_nr in lost_packets {
+                        debug!("SACK: packet {} lost", seq_nr);
+                        self.resend_lost_packet(seq_nr);
+                        packet_loss_detected = true;
+                    }
+                }
+            } else {
+                debug!("Unknown extension {:?}, ignoring", extension.get_type());
             }
+        }
 
-            // Packet lost, halve the congestion window
-            if packet_loss_detected {
-                debug!("packet loss detected, halving congestion window");
-                self.cwnd = max(self.cwnd / 2, MIN_CWND * MAX_DISCV5_PACKET_SIZE);
-                debug!("congestion window: {}", self.cwnd);
-            }
+        // Three duplicate ACKs mean a fast resend request. Resend the first unacknowledged packet
+        // if the incoming packet doesn't have a SACK extension. If it does, the lost packets were
+        // already resent.
+        if !self.send_window.is_empty()
+            && self.duplicate_ack_count == 3
+            && !packet
+                .extensions()
+                .any(|ext| ext.get_type() == ExtensionType::SelectiveAck)
+        {
+            self.resend_lost_packet(packet.ack_nr() + 1);
+        }
 
-            // Success, advance send window
-            self.advance_send_window();
+        // Packet lost, halve the congestion window
+        if packet_loss_detected {
+            debug!("packet loss detected, halving congestion window");
+            self.cwnd = max(self.cwnd / 2, MIN_CWND * MAX_DISCV5_PACKET_SIZE);
+            debug!("congestion window: {}", self.cwnd);
```

Comment:
I'm just curious what your thoughts are on these debug statements. My impression is that once trin is up and running in debug mode, the utp debug comments are going to dominate the majority of the output. Do you agree? Is it worth considering adding another log level just for utp debug output?

Reply:
Probably we need to revisit logging and levels, not sure what is the idiomatic way to do this with the `tracing` library.

Currently, log levels are controlled on a per-module basis, so for example, if we want to run it only for the `portalnet` module we can do it with:

```sh
RUST_LOG=trin_core::portalnet=debug
```
I know this is not the best solution but I need to do more research on rust logging and tracing and come up with a strategy to support multiple log levels, or we can add it as an issue in the repo.